### PR TITLE
Google oauth

### PR DESCRIPTION
**Pull request by isabeltomb**
Adds a new_openid() function to OAuthClient for creating a client for OpenID. Adds an OpenID implementation of the IdentityProvider trait for determining the user's identity. Adds OpenID to the OAuthProvider and the OAuthConfig enums. This is behind a new experimental feature, `oauth-openid`.

Adds the appropriate environment variables and splinterd command line arguments to allow for splinter daemon to build using an OpenID provider for OAuth. Uses the libsplinter `oauth-openid` feature. These changes are behind the existing experimental `auth` feature 

### **Testing:**
Add the following to the docker-compose file
```
    --oauth-client-secret Lj4cszgL9TqJSgfWRZe_LsVf \
    --oauth-client-id 116864774510-j2qt6rchlj85n5k1aj8deiis48akjplt.apps.googleusercontent.com \
    --oauth-redirect-url http://localhost:8080/oauth/callback \
    --oauth-openid-url  https://accounts.google.com/.well-known/openid-configuration \
```
from the root of the repository run 
`CARGO_ARGS="-- --features experimental" docker-compose up --build`

Then go to
`http://localhost:8080/oauth/login?redirect_url=redirect`

### **Testing with splinter admin UI:**
checkout this splinter branch https://github.com/isabeltomb/splinter/tree/test-google-oauth-splinter-admin which has these commits plus a commit to modify splinterd/cargo.toml 

and checkout this splinter-ui branch https://github.com/isabeltomb/splinter-ui/tree/test-oauth-openid which has the necessary docker-compose modifications to run splinter admin ui locally with Google OpenID OAuth 

run `docker-compose -f docker-compose-oauth.yaml up --build` from the root directory of the splinter-ui branch

go to http://localhost:3030

### CODE REVIEW

**Review by vaporos — APPROVED** [review]

File: libsplinter/src/rest_api/mod.rs [hunk 1]
```diff
@@ -957,6 +979,62 @@ impl RestApiBuilder {
...
+            client_secret,
+            discovery_document_response.authorization_endpoint,
+            redirect_url,
+            discovery_document_response.token_endpoint,
+            discovery_document_response.scopes_supported,
+            userinfo_endpoint.clone(),
+        )
+        .map_err(|err| {
+            RestApiServerError::InvalidStateError(InvalidStateError::with_message(format!(
+                "Invalid GitHub OAuth config provided: {}",
```
Comment (peterschwarz) [inline libsplinter/src/rest_api/mod.rs:1020]: s/Github/OpenID

File: libsplinter/src/rest_api/mod.rs [hunk 2]
```diff
@@ -957,6 +979,62 @@ impl RestApiBuilder {
     }
 }
 
+/// Retrieve the OpenId discovery document from the given link
+#[cfg(feature = "oauth-openid")]
+pub fn oauth_client_from_discovery_doc(
```
Comment (peterschwarz) [inline libsplinter/src/rest_api/mod.rs:984]: This function and the struct should probably be under the auth::oauth module.

**Review by peterschwarz — CHANGES_REQUESTED** [review]

File: libsplinter/src/rest_api/mod.rs (hunk 2 above)
Comment (isabeltomb) [reply libsplinter/src/rest_api/mod.rs]: Fixed

File: libsplinter/src/rest_api/mod.rs (hunk 1 above)
Comment (isabeltomb) [reply libsplinter/src/rest_api/mod.rs]: Fixed

**Comment by peterschwarz** [comment]
Tested, works.  The only thing I don't see is a log line for the openid-url config parameter when splinterd starts up in debug (this should be added to `Config::log_as_debug`)

**Review by peterschwarz — DISMISSED** [review]

**Review by shannynalayna — COMMENTED** [review]
There is some information in the splinterd man page around the `oauth-provider` that still needs to updated to say that `openID` is supported, besides just `github`.

**Review by shannynalayna — APPROVED** [review]

**Review by peterschwarz — APPROVED** [review]

**Pull request merged**